### ISSUE TITLE
Installation and initial configuration of CiviCRM on backdropcms.org

### ISSUE THREAD

**Issue by jackaponte** [backdrop-ops/backdropcms.org#789]
Coming out of #782, we plan to install and configure CiviCRM within the Backdrop installation on backdropcms.org. 

During yesterday's meeting @scoobird, @laryn and @larsdesigns volunteered to work together to get that done, with help from others of us as needed. (@ericfg might be interested in helping out on this as well!) Full notes from the meeting are [on this GitHub wiki page.](https://github.com/backdrop-ops/backdropcms.org/wiki/CiviCRM-for-Backdrop-Planning-Meeting-Notes,-April-13-2021)

Morgan (@scoobird) is planning to work on the first steps here; namely, working on a pull request to get CiviCRM code and all needed assets properly incorporated into the GitHub repo for backdropcms.org. (I'd assign this issue to you but you're not in this project yet Morgan, let me know if you need an invite, I forget how that works!)

**Comment by scoobird** [comment]
Hi everyone,

I started working on a PR #791 - to summarize, it currently installs CiviCRM and internationalization files in the contrib modules directory. It also adds a modified version of civicrm.settings.php that is committed to git, and depends on a local version of civicrm.settings.local.php with directory and database connection info, additional settings.

As we discussed it would be great to get some feedback and ideas about how best to coordinate the presence of Civi here to make it least obtrusive for folks working on the site, and minimize the fragility during updates.

Here are some install-related decisions I'd like to get feedback on. @laryn or @ericfg thoughts?
* We sometimes commit the Civi extensions in separate directory because the presence of extension files is required for basic CRM and sometimes CMS functionality. I've left it within www/files so extensions from live would be included in the files when syncing, but alternately can create www/ext and defining that as the extensions directory.
* I've included an example.civicrm.settings.local.php in the repo as a template, and .gitignored all settings.local.php files. Is this suitable or are there other ideas on how to ensure folks can get local settings
* I think to properly install with the PR the first time you have to remove the committed version of civicrm.settings.php then once civi's working, copy the site key and passwords into civicrm.settings.local.php. It would be helpful to get the exact steps and put that into documentation. Or even better if there's a more reliable or less annoying way to do this?
* I haven't dug into how the db sanitization script works, so interested in others' insight in that regard. There are several anonymization scripts for civi we can drop in to preserve data structure, but we could also provide trimmed civicrm tables and/or demo data for development purposes to keep size and risk of sharing private data down.
* As we work though configuration there are other settings we can include in civicrm.settings.php to apply to all sites in all environments.

There are different pre-requisites for A) bare minimum functional site in multiple environments and B) fully functional live site and I'm mostly focusing on A) presently. I also have a working list of step-by-step instructions and pre-requisites to get this functional on the live site, so I can share what I currently have here if folks would find it helpful!

**Comment by larsdesigns** [comment]
@scoobird, This is looks really great. I have an idea to use a database prefix for the tables added by Civi to help with sanitization later. Any thoughts on this?

**Comment by scoobird** [comment]
@larsdesigns adding a custom prefix should work fine. Everything except temp tables will be prefixed by "civicrm_" by default. Let me know what you had in mind -- I presume this should be set across the board in the settings file?

Any other thoughts or feedback on initial PR?

**Comment by jackaponte** [comment]
Hi all--we're picking up work on this again so that we can get Civi installed on backdropcms.org!

@scoobird and I had to hold on this work for a while because of capacity issues at Palante. Morgan didn't hear anything back from @larsdesigns about the initial PR, but we're actually going to submit a new PR soon with Backdrop and CiviCRM fully updated. Morgan will be submitting that new PR after @jenlampton, @stpaultim and I kick the tires on Palante's test site to make sure that the basic Civi installation doesn't break anything or cause any unexpected side effects on the BackdropCMS.org site.

More soon after we've done that testing and gotten that new PR in!

**Comment by jackaponte** [comment]
Hi all! Just checking in on our progress here a week since our last meeting.

@larsdesigns how is setup of the new backdropcms.org beta/test site going? (Also please let me know if there's another issue someplace in GitHub where you're tracking that work!)

A few of us are still double-checking that the Civi install didn't break anything on https://backdropcms.palantetech.coop; I think we'll wrap that up soon, after which we'll get a PR submitted that includes an updated version of CiviCRM that we can try on the new beta/test site on Backdrop infrastructure when it's ready.

**Comment by bugfolder** [comment]
Apologies if this is somewhere and I'm missing it, but one of the tasks before we start populating Civi with data on the live b.org site is a strategy for sanitization of CiviCRM data for people building local versions of b.org from sanitized data/files. If we distribute sanitized Civi dbs, we'll need to be sure to sanitize any custom data fields we create (e.g., Full Name), in addition to Civi's built-in First/Middle/Last, etc.

**Comment by jackaponte** [comment]
> Apologies if this is somewhere and I'm missing it, but one of the tasks before we start populating Civi with data on the live b.org site is a strategy for sanitization of CiviCRM data for people building local versions of b.org from sanitized data/files.

Thanks for raising that, @bugfolder! @jenlampton created a new issue for that, #828, but it wasn't on the [list of tasks above.](https://github.com/backdrop-ops/backdropcms.org/issues/789#issuecomment-928499011) I added that as a task on the list because it's true, we need that in place before we start populating a live b.org instance of Civi with data! I'll also follow up on that separate issue.

**Comment by jackaponte** [comment]
I've submitted pull request #861 for backdropcms.org; it includes CiviCRM 5.44.0!

Our next step here from [this earlier comment:](https://github.com/backdrop-ops/backdropcms.org/issues/789#issuecomment-928499011)

> When the new test/beta site is set up on Backdrop infrastructure, we'll use it to test that new PR, running the CiviCRM installer directly there on the backdropcms.org test site. Again, we'll just be testing the basic CiviCRM installation to make sure it itself doesn't break something on backdropcms.org (which seems unlikely but we want to make sure!)

Coming out of [our meeting earlier this week,](https://github.com/backdrop-ops/backdropcms.org/wiki/CiviCRM-for-BackdropCMS.org-Planning-Meeting-Notes,-December-7-2021) @jenlampton is planning to check in with @larsdesigns about getting beta.backdropcms.org up and running with that PR in place; after that, @bugfolder, @stpaultim, @ericfg and anyone else with the interest will coordinate to test the installation of CiviCRM on the beta site and make sure nothing breaks.

**Comment by jackaponte** [comment]
From @jenlampton on #782 yesterday:

> Hi friends! I think we're going to need some help getting our CiviCRM set up. It feels like it's very close, but I know @larsdesigns has a lot on his plate right now.

Since Justin is busy, is there someone else who can get beta.backdropcms.org up and running with PR #861? After that we can test CiviCRM on the beta site to make sure it doesn't break anything before we get it running on backdropcms.org itself.

**Comment by jackaponte** [comment]
> Since Justin is busy, is there someone else who can get beta.backdropcms.org up and running with PR #861? After that we can test CiviCRM on the beta site to make sure it doesn't break anything before we get it running on backdropcms.org itself.

I just checked #831 and see that it's closed and a note that no other work is needed there, but when I try to visit https://beta.backdropcms.org/ I get a 404 Not Found page.

@jenlampton Do you know what's needed in order to get beta.backdropcms.org up and running with that PR in place, and who might be able to get that done? (Not sure who has the access and skills needed to make that happen!)

**Comment by jenlampton** [comment]
I don't know why beta.backdropcms.org isn't working as it should, I've reopened [that ticket ](https://github.com/backdrop-ops/backdropcms.org/issues/831)with my findings.

That said, I did merge the civi PR into a new branch named `beta` and was able to check out that branch on the beta server, so if we can ever get the server to deliver pages, it should be delivering the right ones :)

**Comment by larsdesigns** [comment]
Completed the following tasks: 

1. Fixed the nginx configuration so that beta.backdropcms.org loads.
2. Updated the beta.backdropcms.org database from production. 
3. I tried to install civicrm using the code provided via git but ran into errors. I found that the civicrm module was installed in the `/modules/contrib/civicrm` directory which was causing the installation errors. According to the CiviCRM installation documentation for Backdrop CMS the path should be `/modules/civicrm`: https://docs.civicrm.org/installation/en/latest/backdrop/. 
4. Downloaded the latest version of CiviCRM for Backdrop CMS and placed it in the `/modules/civicrm` directory. 
5. Performed the installation of CiviCRM in a separate `civicrm` database and updated database user permissions according to the documentation https://docs.civicrm.org/installation/en/latest/general/requirements/. 
6. Updated the mysql credentials file. 
7. Clicked around CiviCRM and it seems to run fine. 

@jenlampton, Please commit the code found in the new `/modules/` path and the configuration files: 
```
	../config/live-active/civicrmtheme.settings.json
	../config/live-active/layout.layout.civicrm_admin_default.json
	modules/civicrm/
```

and delete the `/modules/contrib/civicrm` directory. ` Check the beta.civicrm.org server details. 

Or if it is easier, let me know and I will make the commits. 

I hope the work is acceptable. I did not do anything outside of the CiviCRM documentation. Let me know if anyone has any questions.

**Comment by larsdesigns** [comment]
Also we should probably remove civicrm.settings.php from git repository because it contains credentials.

**Comment by jenlampton** [comment]
> I found that the civicrm module was installed in the /modules/contrib/civicrm directory which was causing the installation errors. ... Please commit the code ...

I've moved the module from `modules/contrib` into `modules` dir. Unfortunately, I'm now getting a ton of errors about `undefined constant CIVICRM_UF` as well as a fatal error because Civi is attempting to write template files in the wrong location: `/home/backdrop/files/civicrm/templates_c": mkdir(): Permission denied`.  (The correct location should be `/home/backdrop/repo/www/files/civicrm`.)

I wonder if the civi settings file we are using is for a different version of civi? 

> and the configuration files:

This is the beta site, so nothing in `live-active` should be changing (or it will affect what's active on the live site, hence the directory name). I have updated the `settings.local.php` file to point to `beta-active` and copied the contents of `live-active` to `beta-active`, committed that to the `beta` branch, and restored everything in `live-active` to the intended state.

I also found a `settings.beta.php` file on the server -- and I deleted it. (settings should be pulled from `settings.local.php`). AFAICT this beta file wasn't being used.

> Clicked around CiviCRM and it seems to run fine.

I was unable to access anything in civicrm because there were no permissions set, so I adjusted permissions so developers can do everything civi-related. These changes were also committed to `beta-active`.

**Comment by laryn** [comment]
I was going to poke around in a spare moment but can't login on the beta site:

> Failed to create "/home/backdrop/files/civicrm/templates_c": mkdir(): Permission denied

**Comment by larsdesigns** [comment]
Uh, our eggs got scrambled here. I will work on resetting it and will update Git repo this time.

**Comment by jenlampton** [comment]
@larsdesigns  the important part of my message above (why the eggs are all scrambled) is this bit:

> Unfortunately, I'm [now getting](https://beta.backdropcms.org/) a ton of errors about undefined constant CIVICRM_UF as well as a fatal error because Civi is attempting to write template files in the wrong location: /home/backdrop/files/civicrm/templates_c": mkdir(): Permission denied. (The correct location should be /home/backdrop/repo/www/files/civicrm.)

I wonder if the civi settings file we are using is for a different version of civi?

**Comment by larsdesigns** [comment]
Completed the following: 

1. Imported new database dump for beta.backdropcms.org. 
2. Updated CiviCRM to 5.46 and committed it to github.com repository. 
3. Performed a fresh install of CiviCRM using the civicrm database to store data. 
4. Reset the passwords for the civicrm and backdrop database users. 
5. Created the `civicrm.settings.php.example` file that is a sanitized copy of the civicrm.setings.php.example file and authored installation instruction at the top of the file. 
6. Updated file permissions for the files directory to make sure civicrm can write to the directory. 
7. Authenticated into the site and browsed around CiviCRM to try and find errors. It seems to be running fine. 
8. Checked Drupal's permissions to make sure the developer role has access to the CiviCRM. 

No further tasks for me at this time. However, I recommend that we setup an automated database backup to be performed a couple times a day while work on Civi.

**Comment by larsdesigns** [comment]
Removed the `civicrm.settings.php` from the git repository and added an entry in .gitignore for it.

**Comment by jackaponte** [comment]
Just catching up here and wow, thanks for all the work on this @larsdesigns and @jenlampton! From the looks of https://beta.backdropcms.org/ we've got an installation up on beta and ready to test.

That relates to one of the tasks set out in my update above:

> * When the new test/beta site is set up on Backdrop infrastructure, we'll use it to test that new PR, running the CiviCRM installer directly there on the backdropcms.org test site. Again, we'll just be testing the basic CiviCRM installation to make sure it itself doesn't break something on backdropcms.org (which seems unlikely but we want to make sure!)

@jenlampton can you confirm that we can rest assured that the just installing CiviCRM itself shouldn't break anything when we move this code to backdropcms.org?

@laryn do you think you could find another spare moment to log into https://beta.backdropcms.org/  and see if anything looks amiss to you on either the Backdrop or the Civi side of things? Does anyone else have time to do that sort of poking around?

After that testing we'll be ready to get this onto the live site, woohoo!

**Comment by jackaponte** [comment]
Questions: who has access to fix things on the beta.backdropcms.org server? How about on the live backdropcms.org server? We'll need to know the latter especially to fix anything server-side that comes up after enabling CiviCRM there, e.g. those warnings that currently appear that https://beta.backdropcms.org/civicrm/a/#/status.

**Comment by jenlampton** [comment]
> @jenlampton can you confirm that we can rest assured that the just installing CiviCRM itself shouldn't break anything when we move this code to backdropcms.org?

Nope! but we can do some testing, and make sure that even if we break everything we can fix it again pretty quickly :)

> Questions: who has access to fix things on the beta.backdropcms.org server?

I think everyone who has access to the live server also has access to beta. @larsdesigns  can you confirm all the keys were copied over?

> How about on the live backdropcms.org server?

myself, @larsdesigns, @bugfolder, @BWPanda, @serundeputy, and @quicksketch I believe.

**Comment by jackaponte** [comment]
Thanks for those responses @jenlampton and @larsdesigns! It's good to know who can be asked for help with things in these two environments.

I logged into the beta site and saw some warnings on the Civi status page that will presumably be fixed right away when CiviCRM is enabled on live. (Wouldn't hurt to fix them on beta too though!)

I did some minimal testing in CiviCRM and nothing seems obviously broken there or in Backdrop, nothing in the watchdog logs. Anyone else want to do more testing on beta.backdropcms.org?

If not, the next task on [our list](https://github.com/backdrop-ops/backdropcms.org/issues/789#issuecomment-928499011) and the only current blocker to enabling CiviCRM on backdropcms.org is this:

> Before enabling Civi on live we need to establish a strategy for sanitization of CiviCRM data for people building local versions of b.org from sanitized data/files (see https://github.com/backdrop-ops/backdropcms.org/issues/828 and https://github.com/backdrop-ops/backdropcms.org/issues/789#issuecomment-968194546)

I'll follow up on that sanitization issue now!

**Comment by jackaponte** [comment]
After [today's planning meeting](https://github.com/backdrop-ops/backdropcms.org/wiki/CiviCRM-for-BackdropCMS.org-Planning-Meeting-Notes,-2022-02-28) we've added some items to our [master to-do list](https://github.com/backdrop-ops/backdropcms.org/issues/789#issuecomment-928499011) and marked some others done. We're getting there!

**Comment by bugfolder** [comment]
Lots of recent progress and updates. See the overall project at https://github.com/backdrop-ops/backdropcms.org/projects/3#card-86238937.

**Comment by bugfolder** [comment]
I think we've either addressed or opened separate issues for everything above. I'm closing this issue in favor of the individual issues [in the CiviCRM project](https://github.com/backdrop-ops/backdropcms.org/projects/3), in the "In Progress" column. They'll get closed out as they're completed.
